internal/mocks/connection.go

```diff
@@ -0,0 +1,298 @@
+	if err != nil {
+		return err
+	}
+	return m.AppendFrame(FrameAMQP, &frames.PerformTransfer{
+		Handle:        linkHandle,
+		DeliveryID:    &deliveryID,
+		DeliveryTag:   []byte("tag"),
+		MessageFormat: &format,
+		Payload:       payloadBuf.Detach(),
+	})
+}
+
+// AppendDisposition appends a PerformDisposition frame with the specified values.
+// The deliveryID MUST match the deliveryID value specified in AppendTransfer.
+func (m *MockConnection) AppendDisposition(deliveryID uint32, state encoding.DeliveryState) error {
+	return m.AppendFrame(FrameAMQP, &frames.PerformDisposition{
+		Role:    encoding.RoleSender,
+		First:   deliveryID,
+		Settled: true,
+		State:   state,
+	})
+}
+
+// AppendFrame adds the specified frame to the list of responses.
+func (m *MockConnection) AppendFrame(t FrameType, f frames.FrameBody) error {
+	raw, err := m.encodeFrame(t, f)
+	if err != nil {
+		return err
+	}
+	m.pendingDataMu.Lock()
+	defer m.pendingDataMu.Unlock()
+	m.pendingData = append(m.pendingData, raw)
+	return nil
+}
+
+func (m *MockConnection) encodeFrame(t FrameType, f frames.FrameBody) ([]byte, error) {
+	bodyBuf := buffer.New([]byte{})
+	if err := encoding.Marshal(bodyBuf, f); err != nil {
+		return nil, err
+	}
+	// create the frame header, needs size of the body plus itself
+	header := frameHeader{
+		Size:       uint32(bodyBuf.Len()) + 8,
+		DataOffset: 2,
+		FrameType:  uint8(t),
+	}
+	headerBuf := buffer.New([]byte{})
+	if err := encoding.Marshal(headerBuf, header); err != nil {
+		return nil, err
+	}
+	// concatenate header + body
+	raw := headerBuf.Detach()
+	raw = append(raw, bodyBuf.Detach()...)
+	return raw, nil
+}
+
+///////////////////////////////////////////////////////
+// following functions are for the net.Conn interface
+///////////////////////////////////////////////////////
+
+func (m *MockConnection) Read(b []byte) (n int, err error) {
+	if err := m.closedCheck(); err != nil {
+		return 0, err
+	}
+	select {
+	case <-m.readDL.C:
+		return 0, errors.New("read deadline exceeded")
+	case rd := <-m.readData:
+		if !m.readDL.Stop() {
+			<-m.readDL.C
+		}
+		m.readDL.Reset(m.readDLDur)
+		return copy(b, rd), nil
+	}
+}
+
+func (m *MockConnection) Write(b []byte) (n int, err error) {
+	if err := m.closedCheck(); err != nil {
+		return 0, err
+	}
+	if m.FrameCallback != nil {
+		frame, err := decodeFrame(b)
+		if err != nil {
+			return 0, err
+		}
+		if err := m.FrameCallback(frame); err != nil {
+			return 0, err
+		}
+	}
+	m.pendingDataMu.Lock()
+	if len(m.pendingData) == 0 {
+		m.pendingDataMu.Unlock()
+		return 0, errors.New("no more responses")
+	}
+	// pop the front of pending data and send it to the read channel
+	out := m.pendingData[0]
+	m.pendingData = m.pendingData[1:]
+	m.pendingDataMu.Unlock()
+	m.readData <- out
+	// swallow the write
+	return len(b), nil
+}
+
+func (m *MockConnection) Close() error {
+	m.closedMu.Lock()
+	defer m.closedMu.Unlock()
+	m.closed = true
+	if !m.readDL.Stop() {
```

Comment:
I feel like it'd be useful to have a check that says "you shouldn't have anything left in the .readDL channel".

This might not be true in real world usage but in our tests I'm assuming it'd be a bug for us to leave behind any data.

Reply:
Great idea!